File: client/cli/src/handshake/handshakeHandler.ts
/**
 * Authenticated key-exchange handshake handler.
 *
 * Orchestrates the 3-step Noise-XK-inspired handshake using @aicq/crypto
 * primitives and the server's relay API.
 */

import type { KeyPair, HandshakeRequest, HandshakeResponse } from '@aicq/crypto';
import {
  generateKeyExchangeKeyPair,
  getPublicKeyFingerprint,
  createHandshakeRequest,
  createHandshakeResponse,
  completeHandshake,
  encodeBase64,
  decodeBase64,
} from '@aicq/crypto';
import { APIClient } from '../services/apiClient.js';
import { WSClient } from '../services/wsClient.js';
import { IdentityManager } from '../services/identityManager.js';
import { ClientStore } from '../store.js';
import type { HandshakeProgress, FriendInfo } from '../types.js';

/** Incoming handshake request waiting for user action. */
interface PendingIncomingHandshake {
  sessionId: string;
  requesterId: string;
  requesterPublicKey: string;
  request: HandshakeRequest;
  timestamp: number;
}

export class HandshakeHandler {
  private api: APIClient;
  private ws: WSClient;
  private identity: IdentityManager;
  private store: ClientStore;

  /** Active outgoing handshake ephemeral keys, keyed by sessionId. */
  private outgoingEphemeral = new Map<
    string,
    { keys: KeyPair; request: HandshakeRequest; targetNodeId: string }
  >();

  /** Incoming handshake requests awaiting user accept/reject. */
  private incomingRequests = new Map<string, PendingIncomingHandshake>();

  /** Progress callback set by the caller. */
  private progressCallback: ((progress: HandshakeProgress) => void) | null = null;

  constructor(
    api: APIClient,
    ws: WSClient,
    identity: IdentityManager,
    store: ClientStore,
  ) {
    this.api = api;
    this.ws = ws;
    this.identity = identity;
    this.store = store;

    this._setupListeners();
  }

  /* ──────────────── Progress callback ──────────────── */

  onProgress(callback: (progress: HandshakeProgress) => void): void {
    this.progressCallback = callback;
  }

  private emitProgress(progress: HandshakeProgress): void {
    this.progressCallback?.(progress);
  }

  /* ──────────────── WebSocket listeners ──────────────── */

  private _setupListeners(): void {
    // Listen for incoming handshake signals relayed via the server
    this.ws.on('signal', (msg: any) => {
      if (msg.data?.type === 'handshake_request') {
        this._handleIncomingRequest(msg.data);
      } else if (msg.data?.type === 'handshake_response') {
        this._handleResponse(msg.data);
      } else if (msg.data?.type === 'handshake_confirm') {
        this._handleConfirm(msg.data);
      }
    });
  }

  /* ──────────────── Initiate (outgoing) ──────────────── */

  /**
   * Initiate a handshake with a peer identified by their temp number.
   *
   * Flow:
   *  1. Resolve temp number → get targetNodeId
   *  2. Generate ephemeral X25519 key pair
   *  3. Create handshake request (identity PK + ephemeral PK)
   *  4. Submit to server via API
   *  5. Send handshake_request signal to target via WebSocket
   *  6. Wait for response
   *  7. Process response, derive session key, send confirm
   *  8. Wait for confirm ack
   *  9. Store session key, add friend
   */
  async initiateHandshake(tempNumber: string): Promise<void> {
    // Step 1: Resolve temp number
    this.emitProgress({ status: 'initiating', detail: 'Resolving temp number...' });
    let targetNodeId: string;
    try {
      const resolved = await this.api.resolveTempNumber(tempNumber);
      targetNodeId = resolved.nodeId;
    } catch (err) {
      this.emitProgress({ status: 'failed', detail: `Temp number resolution failed: ${err}` });
      throw err;
    }

    // Step 2: Generate ephemeral key pair
    const ephemeralKeys = generateKeyExchangeKeyPair();

    // Step 3: Create handshake request
    const request: HandshakeRequest = {
      identityPublicKey: this.identity.getPublicKey(),
      ephemeralPublicKey: ephemeralKeys.publicKey,
    };

    // Step 4: Submit to server
    this.emitProgress({ status: 'initiating', detail: 'Starting handshake session...' });
    let sessionId: string;
    try {
      const result = await this.api.initiateHandshake(tempNumber);
      sessionId = result.sessionId;
    } catch (err) {
      this.emitProgress({ status: 'failed', detail: `Failed to initiate handshake: ${err}` });
      throw err;
    }

    // Store ephemeral keys for later
    this.outgoingEphemeral.set(sessionId, {
      keys: ephemeralKeys,
      request,
      targetNodeId,
    });

    // Step 5: Send signal to target via WebSocket
    this.ws.send('signal', {
      to: targetNodeId,
      data: {
        type: 'handshake_request',
        sessionId,
        identityPublicKey: encodeBase64(request.identityPublicKey),
        ephemeralPublicKey: encodeBase64(request.ephemeralPublicKey),
      },
    });

    // Step 6: Wait for response via WebSocket
    this.emitProgress({
      status: 'waiting_response',
      detail: 'Waiting for peer to accept...',
      peerInfo: { id: targetNodeId, publicKey: '', fingerprint: '' },
    });

    try {
      const responseMsg = await this.ws.waitForResponse(
        sessionId,
        'signal',
        120_000, // 2 minute timeout
      );

      if (responseMsg.data?.type !== 'handshake_response') {
        throw new Error('Unexpected response type');
      }

      // Step 7: Process response, derive session key
      this.emitProgress({ status: 'processing_response', detail: 'Processing response...' });
      await this._completeOutgoingHandshake(sessionId, responseMsg.data);
    } catch (err) {
      this.outgoingEphemeral.delete(sessionId);
      this.emitProgress({ status: 'failed', detail: `Handshake failed: ${err}` });
      throw err;
    }
  }

  /**
   * Complete the outgoing handshake after receiving the response.
   */
  private async _completeOutgoingHandshake(
    sessionId: string,
    responseData: any,
  ): Promise<void> {
    const state = this.outgoingEphemeral.get(sessionId);
    if (!state) throw new Error('No outgoing handshake state found');

    const targetIdentityPubKey = decodeBase64(responseData.identityPublicKey);
    const targetEphemeralPubKey = decodeBase64(responseData.ephemeralPublicKey);
    const proof = decodeBase64(responseData.proof);

    const response: HandshakeResponse = {
      identityPublicKey: targetIdentityPubKey,
      ephemeralPublicKey: targetEphemeralPubKey,
      proof,
    };

    // Derive session key and verify proof
    const myIdentityKeys: KeyPair = {
      publicKey: this.identity.getPublicKey(),
      secretKey: this.identity.getSigningSecretKey(),
    };

    const sessionKey = completeHandshake(
      response,
      state.request,
      myIdentityKeys,
      state.keys,
    );

    // Store session key
    this.store.setSessionKey(state.targetNodeId, sessionKey);

    // Submit confirm to server API
    this.emitProgress({ status: 'confirming', detail: 'Confirming handshake...' });
    // We send a simple confirm signal back — the proof of successful
    // key derivation is enough.  For the server we submit a minimal confirm.
    await this.api.submitHandshakeConfirm(
      sessionId,
      Buffer.from(sessionKey.slice(0, 16)), // truncated proof of derivation
    );

    // Send confirm to peer via WebSocket
    this.ws.send('signal', {
      to: state.targetNodeId,
      data: {
        type: 'handshake_confirm',
        sessionId,
      },
    });

    // Add friend to store
    const friendInfo: FriendInfo = {
      id: state.targetNodeId,
      publicKey: encodeBase64(targetIdentityPubKey),
      fingerprint: getPublicKeyFingerprint(targetIdentityPubKey),
      addedAt: new Date().toISOString(),
      lastSeen: new Date().toISOString(),
      isOnline: false,
    };
    this.store.addFriend(friendInfo);
    this.store.save();

    // Cleanup
    this.outgoingEphemeral.delete(sessionId);

    this.emitProgress({
      status: 'completed',
      detail: `Handshake complete with ${state.targetNodeId}`,
      peerInfo: {
        id: state.targetNodeId,
        publicKey: encodeBase64(targetIdentityPubKey),
        fingerprint: friendInfo.fingerprint,
      },
    });
  }

  /* ──────────────── Incoming handshake ──────────────── */

  /**
   * Called when an incoming handshake request arrives via WebSocket.
   * Stores the request and notifies the user via progress callback.
   */
  private _handleIncomingRequest(data: any): void {
    const sessionId: string = data.sessionId;
    const requesterId: string = data.requesterId;
    const identityPubKey = decodeBase64(data.identityPublicKey);
    const ephemeralPubKey = decodeBase64(data.ephemeralPublicKey);

    const request: HandshakeRequest = {
      identityPublicKey: identityPubKey,
      ephemeralPublicKey: ephemeralPubKey,
    };

    this.incomingRequests.set(sessionId, {
      sessionId,
      requesterId,
      requesterPublicKey: encodeBase64(identityPubKey),
      request,
      timestamp: Date.now(),
    });

    this.emitProgress({
      status: 'incoming_request',
      detail: `Incoming handshake from ${requesterId.slice(0, 8)}...`,
      peerInfo: {
        id: requesterId,
        publicKey: encodeBase64(identityPubKey),
        fingerprint: getPublicKeyFingerprint(identityPubKey),
      },
    });
  }

  /* ──────────────── Accept / Reject ──────────────── */

  /**
   * Accept an incoming handshake request.
   */
  async acceptHandshake(requestId: string): Promise<void> {
    const pending = this.incomingRequests.get(requestId);
    if (!pending) {
      throw new Error('No pending handshake request found');
    }

    // Generate ephemeral key pair
    const ephemeralKeys = generateKeyExchangeKeyPair();

    // Create response
    const myIdentityKeys: KeyPair = {
      publicKey: this.identity.getPublicKey(),
      secretKey: this.identity.getSigningSecretKey(),
    };

    const response = createHandshakeResponse(
      pending.request,
      myIdentityKeys,
      ephemeralKeys,
    );

    // Derive session key (responder side) using the same function
    // that the initiator will use, ensuring both sides agree.
    const sessionKey = completeHandshake(
      response,
      pending.request,
      myIdentityKeys,
      ephemeralKeys,
    );

    // Store session key
    this.store.setSessionKey(pending.requesterId, sessionKey);

    // Submit response to server API
    const responseBuf = Buffer.from(
      JSON.stringify({
        identityPublicKey: encodeBase64(response.identityPublicKey),
        ephemeralPublicKey: encodeBase64(response.ephemeralPublicKey),
        proof: encodeBase64(response.proof),
      }),
      'utf-8',
    );
    await this.api.submitHandshakeResponse(pending.sessionId, responseBuf);

    // Send response to requester via WebSocket
    this.ws.send('signal', {
      to: pending.requesterId,
      data: {
        type: 'handshake_response',
        sessionId: pending.sessionId,
        identityPublicKey: encodeBase64(response.identityPublicKey),
        ephemeralPublicKey: encodeBase64(response.ephemeralPublicKey),
        proof: encodeBase64(response.proof),
      },
    });

    // Wait for confirm from initiator
    this.emitProgress({
      status: 'confirming',
      detail: 'Waiting for confirmation...',
    });

    try {
      await this.ws.waitForResponse(pending.sessionId, 'signal', 60_000);
    } catch (err) {
      console.warn('[Handshake] Confirm timeout:', err);
    }

    // Add friend
    const friendInfo: FriendInfo = {
      id: pending.requesterId,
      publicKey: pending.requesterPublicKey,
      fingerprint: getPublicKeyFingerprint(pending.request.identityPublicKey),
      addedAt: new Date().toISOString(),
      lastSeen: new Date().toISOString(),
      isOnline: true,
    };
    this.store.addFriend(friendInfo);
    this.store.save();

    this.incomingRequests.delete(requestId);

    this.emitProgress({
      status: 'completed',
      detail: `Friend ${pending.requesterId.slice(0, 8)}... added`,
      peerInfo: {
        id: pending.requesterId,
        publicKey: pending.requesterPublicKey,
        fingerprint: friendInfo.fingerprint,
      },
    });
  }

  /**
   * Reject an incoming handshake request.
   */
  rejectHandshake(requestId: string): void {
    const pending = this.incomingRequests.get(requestId);
    if (!pending) return;

    // Notify the requester
    this.ws.send('signal', {
      to: pending.requesterId,
      data: {
        type: 'handshake_rejected',
        sessionId: pending.sessionId,
      },
    });

    this.incomingRequests.delete(requestId);

    this.emitProgress({
      status: 'rejected',
      detail: 'Handshake request rejected',
    });
  }

  /* ──────────────── Response / Confirm handlers (outgoing) ──────────────── */

  private _handleResponse(_data: any): void {
    // The waitForResponse mechanism in WSClient handles response routing.
    // No buffering needed here.
  }

  private _handleConfirm(_data: any): void {
    // The waitForResponse mechanism in WSClient handles confirm routing.
    // No buffering needed here.
  }

  /* ──────────────── Session key lookup ──────────────── */

  /** Retrieve the session key for a given peer. */
  getSessionKey(peerId: string): Uint8Array | null {
    return this.store.getSessionKey(peerId);
  }

  /* ──────────────── Cleanup ──────────────── */

  destroy(): void {
    this.outgoingEphemeral.clear();
    this.incomingRequests.clear();
    this.progressCallback = null;
  }
}

File: client/handshake_handler.py
"""
Authenticated key-exchange handshake handler.

Orchestrates the 3-step Noise-XK-inspired handshake using the shared
crypto library primitives and the server's relay API.
"""

from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional, Callable

import sys
_shared = str(Path(__file__).resolve().parent.parent / "shared")
if _shared not in sys.path:
    sys.path.insert(0, _shared)

from crypto import (
    KeyPair,
    HandshakeRequest,
    HandshakeResponse,
    generate_key_exchange_keypair,
    get_public_key_fingerprint,
    create_handshake_request,
    create_handshake_response,
    complete_handshake,
    encode_base64,
    decode_base64,
)

from .api_client import APIClient
from .ws_client import WSClient
from .identity_manager import IdentityManager
from .db import ClientDatabase
from .types import HandshakeProgress, HandshakeStatus, FriendInfo, FriendType


class HandshakeHandler:
    """Handles the 3-step authenticated key-exchange handshake.

    Supports both initiator and responder roles. The initiator starts
    a handshake by resolving a temp number, while the responder receives
    incoming requests and accepts or rejects them.
    """

    def __init__(
        self,
        api: APIClient,
        ws: WSClient,
        identity: IdentityManager,
        db: ClientDatabase,
    ):
        self._api = api
        self._ws = ws
        self._identity = identity
        self._db = db

        # Active outgoing handshake ephemeral keys, keyed by session_id
        self._outgoing: dict[str, dict] = {}

        # Incoming handshake requests awaiting user accept/reject
        self._incoming: dict[str, dict] = {}

        # Progress callback
        self._progress_cb: Optional[Callable[[HandshakeProgress], None]] = None

        # Setup WebSocket listeners
        self._ws.on("signal", self._on_signal)

    def on_progress(self, callback: Callable[[HandshakeProgress], None]) -> None:
        self._progress_cb = callback

    def _emit_progress(self, progress: HandshakeProgress) -> None:
        if self._progress_cb:
            try:
                self._progress_cb(progress)
            except Exception:
                pass

    # ──────────────── WebSocket signal handler ────────────────

    async def _on_signal(self, msg: dict) -> None:
        data = msg.get("data", {})
        signal_type = data.get("type", "")

        if signal_type == "handshake_request":
            await self._handle_incoming_request(data)
        elif signal_type == "handshake_response":
            await self._handle_response(data)
        elif signal_type == "handshake_confirm":
            await self._handle_confirm(data)

    # ──────────────── Initiate (outgoing) ────────────────

    async def initiate_handshake(self, temp_number: str) -> None:
        """Initiate a handshake with a peer identified by temp number.

        Flow:
            1. Resolve temp number -> target node ID
            2. Generate ephemeral X25519 key pair
            3. Create handshake request
            4. Submit to server via API
            5. Send handshake_request signal via WebSocket
            6. Wait for response
            7. Process response, derive session key, send confirm
            8. Store session key, add friend
        """
        # Step 1: Resolve temp number
        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.INITIATING,
            detail="Resolving temp number...",
        ))
        try:
            resolved = await self._api.resolve_temp_number(temp_number)
            target_node_id = resolved["nodeId"]
        except Exception as exc:
            self._emit_progress(HandshakeProgress(
                status=HandshakeStatus.FAILED,
                detail=f"Temp number resolution failed: {exc}",
            ))
            raise

        # Step 2-3: Generate ephemeral keys and create request
        ephemeral_keys = generate_key_exchange_keypair()
        request = HandshakeRequest(
            identity_public_key=self._identity.get_public_key(),
            ephemeral_public_key=ephemeral_keys.public_key,
        )

        # Step 4: Submit to server
        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.INITIATING,
            detail="Starting handshake session...",
        ))
        try:
            result = await self._api.initiate_handshake(temp_number)
            session_id = result["sessionId"]
        except Exception as exc:
            self._emit_progress(HandshakeProgress(
                status=HandshakeStatus.FAILED,
                detail=f"Failed to initiate handshake: {exc}",
            ))
            raise

        # Store ephemeral keys for later
        self._outgoing[session_id] = {
            "keys": ephemeral_keys,
            "request": request,
            "target_node_id": target_node_id,
        }

        # Step 5: Send signal via WebSocket
        self._ws.send("signal", {
            "to": target_node_id,
            "data": {
                "type": "handshake_request",
                "sessionId": session_id,
                "identityPublicKey": encode_base64(request.identity_public_key),
                "ephemeralPublicKey": encode_base64(request.ephemeral_public_key),
            },
        })

        # Step 6: Wait for response
        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.WAITING_RESPONSE,
            detail="Waiting for peer to accept...",
            peer_info=FriendInfo(id=target_node_id),
        ))

        try:
            response_msg = await self._ws.wait_for_response(
                session_id, "signal", timeout=120.0
            )
            resp_data = response_msg.get("data", {})
            if resp_data.get("type") != "handshake_response":
                raise RuntimeError("Unexpected response type")

            await self._complete_outgoing_handshake(session_id, resp_data)
        except Exception as exc:
            self._outgoing.pop(session_id, None)
            self._emit_progress(HandshakeProgress(
                status=HandshakeStatus.FAILED,
                detail=f"Handshake failed: {exc}",
            ))
            raise

    async def _complete_outgoing_handshake(
        self, session_id: str, response_data: dict
    ) -> None:
        """Complete the outgoing handshake after receiving the response."""
        state = self._outgoing.get(session_id)
        if not state:
            raise RuntimeError("No outgoing handshake state found")

        target_identity_pub = decode_base64(response_data["identityPublicKey"])
        target_ephemeral_pub = decode_base64(response_data["ephemeralPublicKey"])
        proof = decode_base64(response_data["proof"])

        response = HandshakeResponse(
            identity_public_key=target_identity_pub,
            ephemeral_public_key=target_ephemeral_pub,
            proof=proof,
        )

        my_identity_keys = KeyPair(
            public_key=self._identity.get_public_key(),
            secret_key=self._identity.get_signing_secret_key(),
        )

        session_key = complete_handshake(
            response, state["request"], my_identity_keys, state["keys"]
        )

        # Store session key
        target_id = state["target_node_id"]
        await self._db.set_session_key(target_id, session_key)

        # Submit confirm to server
        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.CONFIRMING,
            detail="Confirming handshake...",
        ))
        await self._api.submit_handshake_confirm(
            session_id, session_key[:16]
        )

        # Send confirm signal
        self._ws.send("signal", {
            "to": target_id,
            "data": {"type": "handshake_confirm", "sessionId": session_id},
        })

        # Add friend
        friend = FriendInfo(
            id=target_id,
            public_key=encode_base64(target_identity_pub),
            fingerprint=get_public_key_fingerprint(target_identity_pub),
            added_at=datetime.now(tz=timezone.utc).isoformat(),
            last_seen=datetime.now(tz=timezone.utc).isoformat(),
            is_online=False,
            friend_type=FriendType.HUMAN,
        )
        await self._db.add_friend(friend)

        self._outgoing.pop(session_id, None)
        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.COMPLETED,
            detail=f"Handshake complete with {target_id[:8]}...",
            peer_info=friend,
        ))

    # ──────────────── Incoming handshake ────────────────

    async def _handle_incoming_request(self, data: dict) -> None:
        session_id = data.get("sessionId", "")
        requester_id = data.get("requesterId", "")
        identity_pub = decode_base64(data.get("identityPublicKey", ""))
        ephemeral_pub = decode_base64(data.get("ephemeralPublicKey", ""))

        request = HandshakeRequest(
            identity_public_key=identity_pub,
            ephemeral_public_key=ephemeral_pub,
        )

        self._incoming[session_id] = {
            "session_id": session_id,
            "requester_id": requester_id,
            "requester_public_key": encode_base64(identity_pub),
            "request": request,
            "timestamp": datetime.now(tz=timezone.utc).isoformat(),
        }

        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.WAITING_RESPONSE,
            detail=f"Incoming handshake from {requester_id[:8]}...",
            peer_info=FriendInfo(
                id=requester_id,
                public_key=encode_base64(identity_pub),
                fingerprint=get_public_key_fingerprint(identity_pub),
            ),
        ))

    # ──────────────── Accept / Reject ────────────────

    async def accept_handshake(self, request_id: str) -> None:
        """Accept an incoming handshake request."""
        pending = self._incoming.get(request_id)
        if not pending:
            raise RuntimeError("No pending handshake request found")

        ephemeral_keys = generate_key_exchange_keypair()

        my_identity_keys = KeyPair(
            public_key=self._identity.get_public_key(),
            secret_key=self._identity.get_signing_secret_key(),
        )

        response, session_key = create_handshake_response(
            pending["request"], my_identity_keys, ephemeral_keys
        )

        await self._db.set_session_key(pending["requester_id"], session_key)

        # Submit response to server API
        import json
        response_buf = json.dumps({
            "identityPublicKey": encode_base64(response.identity_public_key),
            "ephemeralPublicKey": encode_base64(response.ephemeral_public_key),
            "proof": encode_base64(response.proof),
        }).encode("utf-8")
        await self._api.submit_handshake_response(pending["session_id"], response_buf)

        # Send response signal
        self._ws.send("signal", {
            "to": pending["requester_id"],
            "data": {
                "type": "handshake_response",
                "sessionId": pending["session_id"],
                "identityPublicKey": encode_base64(response.identity_public_key),
                "ephemeralPublicKey": encode_base64(response.ephemeral_public_key),
                "proof": encode_base64(response.proof),
            },
        })

        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.CONFIRMING,
            detail="Waiting for confirmation...",
        ))

        try:
            await self._ws.wait_for_response(pending["session_id"], "signal", timeout=60.0)
        except Exception:
            pass  # Confirm timeout is non-fatal

        friend = FriendInfo(
            id=pending["requester_id"],
            public_key=pending["requester_public_key"],
            fingerprint=get_public_key_fingerprint(pending["request"].identity_public_key),
            added_at=datetime.now(tz=timezone.utc).isoformat(),
            last_seen=datetime.now(tz=timezone.utc).isoformat(),
            is_online=True,
        )
        await self._db.add_friend(friend)
        self._incoming.pop(request_id, None)

        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.COMPLETED,
            detail=f"Friend {pending['requester_id'][:8]}... added",
            peer_info=friend,
        ))

    def reject_handshake(self, request_id: str) -> None:
        """Reject an incoming handshake request."""
        pending = self._incoming.get(request_id)
        if not pending:
            return

        self._ws.send("signal", {
            "to": pending["requester_id"],
            "data": {
                "type": "handshake_rejected",
                "sessionId": pending["session_id"],
            },
        })
        self._incoming.pop(request_id, None)

        self._emit_progress(HandshakeProgress(
            status=HandshakeStatus.REJECTED,
            detail="Handshake request rejected",
        ))

    async def get_session_key(self, peer_id: str) -> Optional[bytes]:
        return await self._db.get_session_key(peer_id)

    # ──────────────── Internal handlers ────────────────

    async def _handle_response(self, data: dict) -> None:
        # Handled via waitForResponse mechanism
        pass

    async def _handle_confirm(self, data: dict) -> None:
        # Handled via waitForResponse mechanism
        pass

    # ──────────────── Cleanup ────────────────

    def destroy(self) -> None:
        self._outgoing.clear()
        self._incoming.clear()
        self._progress_cb = None
